### PR TITLE
add pre-commits and some minor changes for contributors

### PR DESCRIPTION
**Pull request by HamedBabaei**
Hey @arxyzan and @pooya-mohammadi 

I hope this finds you well and thanks for such a great open-source tool!

I have added pre-commits to the repository to ease contributions and automation of developments. I also run it on all the existing files in this repository so you may find changes in files. in summary, the changes are:
- add pre-commit & its changes 
- add navigation badges to the readme
- add contribution guideline 

I highly recommend adding the `requirements.txt` file to the repository beside the `.toml` file

Keep up the good work.
Best Regards,
Hamed

### CODE REVIEW

**Comment by arxyzan** [comment]
Hello @HamedBabaei jan and thanks for diving into the project and having a pull request sent already!
As of now, Hezar really lacks well-defined tests specially for pull requests and unfortunately, we have to review the changes manually. 
This PR seems pretty huge. Although the changes are minor but they are present in most files. The changes for pre-commit and docs cleanups are solid and can be merged for sure.
But I noticed that every Python file has a `# -*- coding: utf-8 -*-` inserted on top which is unnecessary for Python 3 since the default encoding is actually utf-8. Can you propose a workaround to disable this?

File: hezar/__init__.py
Comment (arxyzan) [inline hezar/__init__.py:1]: @HamedBabaei Please don't reorder imports on this file as it may cause cyclic imports right now. specially the registry which must come first.

**Comment by HamedBabaei** [comment]
Hello @arxyzan 

Thanks for the quick review. 

I was adding those encodings based on common practices that I was following in my projects, so I just removed them here and fixed the pre-commit about it. Also made some minor modifications for the variable name that pre-commit was getting an issue (e.g. `l` changed to `label`) and some other minor fixes!

If I get some time I will try to add some test cases to ease the testing process as well!

File: hezar/__init__.py
Comment (HamedBabaei) [reply hezar/__init__.py:1]: Done! I roled back changes for this file and ignored sort for `.registery` 

thanks for the hint!

**Comment by HamedBabaei** [comment]
> @HamedBabaei I forgot to mention Hamed jan, I don't know what linting rules you've used in pre-commit but we always wrap long lines in the paranthesis, brackets, etc to the next line. This applies to all function, class definitions, data structures, and imports.
> 
> ```python
> # WRONG
> from .configs import (DatasetConfig, EmbeddingConfig, MetricConfig,
>                       ModelConfig, PreprocessorConfig)
> 
> # CORRECT
> from .configs import (
>     DatasetConfig,
>     EmbeddingConfig,
>     MetricConfig,
>     ModelConfig,
>     PreprocessorConfig,
> )
> ```
> 
> ```python
> # WRONG
> def __init__(self,
>              model: Model = None,
>              config: TrainerConfig = None,
>              train_dataset: Dataset = None,
>              eval_dataset: Dataset = None,
>              data_collator: Callable = None,
>              preprocessor: Union[Preprocessor, PreprocessorsContainer] = None,
>              optimizer: torch.optim.Optimizer = None,
>              lr_scheduler=None,
>              **kwargs,
>              ):
> 
> # CORRECT
> def __init__(
>     self,
>     model: Model = None,
>     config: TrainerConfig = None,
>     train_dataset: Dataset = None,
>     eval_dataset: Dataset = None,
>     data_collator: Callable = None,
>     preprocessor: Union[Preprocessor, PreprocessorsContainer] = None,
>     optimizer: torch.optim.Optimizer = None,
>     lr_scheduler=None,
>     **kwargs,
> ):
> ```

@arxyzan 

I see, I will check out the `flake8` to see how I can fix it to work the same way you have it. it seems that you are using `C,E,F,I,W` linting configurations in `.toml` file and I added this (+ other configurations) to the flake (with black configs) as well, But again the desired output only generated for imports lines that exceed 120 character limits. The codes that exceed this limit should be dropped into multiple lines and if it is not, then it should be in one line in the way that you suggested (I should do the fix for this but before this I need confirmation).

for example in the builders.py script, the following lines do not exceed the 120-character limit, so they should be in one line? or multiple lines?
```python
from .configs import DatasetConfig, EmbeddingConfig, MetricConfig, ModelConfig, PreprocessorConfig

from .registry import datasets_registry, embeddings_registry, metrics_registry, models_registry, preprocessors_registry
```

File: hezar/embeddings/embedding.py [hunk 1]
```diff
@@ -26,19 +25,20 @@
 
 # Check if the right combo of gensim/numpy versions are installed
 def _verify_gensim_installation():
-    if (
```
Comment (arxyzan) [inline hezar/embeddings/embedding.py:29]: @HamedBabaei It's better to keep this unchanged as it is for better readability.

File: hezar/models/__init__.py [hunk 2]
```diff
@@ -1,10 +1,10 @@
-from .model import Model
```
Comment (arxyzan) [inline hezar/models/__init__.py:1]: @HamedBabaei jan, base classes must be on top of the `__init__` files. You can ignore sorting here.

File: hezar/preprocessors/__init__.py [hunk 3]
```diff
@@ -1,5 +1,5 @@
-from .preprocessor import Preprocessor, PreprocessorsContainer
```
Comment (arxyzan) [inline hezar/preprocessors/__init__.py:1]: @HamedBabaei Same here. base class comes first.

File: hezar/preprocessors/preprocessor.py [hunk 4]
```diff
@@ -82,10 +77,7 @@ def load(
                     config_file = os.path.join(hub_or_local_path, subfolder, f)
                 else:
                     config_file = hf_hub_download(
-                        hub_or_local_path,
```
Comment (arxyzan) [inline hezar/preprocessors/preprocessor.py:85]: @HamedBabaei  How does this not conform to the wrapping rules?

File: hezar/preprocessors/preprocessor.py [hunk 5]
```diff
@@ -49,13 +50,7 @@ def push_to_hub(
         raise NotImplementedError
 
     @classmethod
-    def load(
```
Comment (arxyzan) [inline hezar/preprocessors/preprocessor.py:52]: @HamedBabaei jan, such core methods like `load`, `save`, `push_to_hub` function defenitions might not exceed the 120 characters limit, but for readability, we prefer to wrap them. Is there any solution to keep such cases untouched?

File: hezar/preprocessors/text_normalizer.py [hunk 6]
```diff
@@ -79,13 +83,7 @@ def __call__(
         return inputs
 
     @classmethod
-    def load(
```
Comment (arxyzan) [inline hezar/preprocessors/text_normalizer.py:82]: @HamedBabaei same here.

File: hezar/trainers/trainer.py [hunk 7]
```diff
@@ -511,11 +511,7 @@ def push_to_hub(
             commit_message=commit_message,
         )
         self.train_dataset.config.push_to_hub(
-            repo_id,
```
Comment (arxyzan) [inline hezar/trainers/trainer.py:514]: @HamedBabaei jan, what exactly is causing this behavior? Shouldn't all parameters be on one line each?

File: hezar/embeddings/embedding.py (hunk 1 above)
Comment (HamedBabaei) [reply hezar/embeddings/embedding.py:29]: @arxyzan I see, this is done now!

File: hezar/models/__init__.py (hunk 2 above)
Comment (HamedBabaei) [reply hezar/models/__init__.py:1]: Hey @arxyzan jan, is there any reason for this? if this is a personal preference of having base classes on top, we might need to use the `# isort: skip` command for that specific import. If this is not we might consider ignoring sorting in `__init__` files!

Comment (arxyzan) [reply hezar/models/__init__.py:1]: The main reason is that we faced some issues with cyclic imports in the past due to this. Having the base class imported on top of the `__init__` file solves the issue.

File: hezar/preprocessors/preprocessor.py (hunk 4 above)
Comment (HamedBabaei) [reply hezar/preprocessors/preprocessor.py:85]: @arxyzan nice catch, I checked out this and it seems to be black was causing the issue, so I fixed it manually (and tested to make sure it will not change it) for this and the rest of the catch that you provided, and it works now!

File: hezar/preprocessors/text_normalizer.py (hunk 6 above)
Comment (HamedBabaei) [reply hezar/preprocessors/text_normalizer.py:82]: done!

File: hezar/preprocessors/preprocessor.py (hunk 4 above)
Comment (arxyzan) [reply hezar/preprocessors/preprocessor.py:85]: @HamedBabaei  Thanks for your efforts <3

File: hezar/preprocessors/text_normalizer.py (hunk 6 above)
Comment (arxyzan) [reply hezar/preprocessors/text_normalizer.py:82]: @HamedBabaei Brilliant!

File: hezar/trainers/trainer.py (hunk 7 above)
Comment (HamedBabaei) [reply hezar/trainers/trainer.py:514]: @arxyzan Nice hint! The way that you put this before was right and in the middle of the way we configured it to consider multiple lines for parameters, so this change wasn't reverted back somehow, so I reverted back manually and runner pre-commit to make sure it works! 

Thanks for pointing this out!

File: hezar/models/__init__.py (hunk 2 above)
Comment (HamedBabaei) [reply hezar/models/__init__.py:1]: I see, then I just put `# isort: skip` for base classes, it will make it easier to follow up. thanks!

File: hezar/preprocessors/__init__.py (hunk 3 above)
Comment (HamedBabaei) [reply hezar/preprocessors/__init__.py:1]: done!

File: hezar/preprocessors/tokenizers/__init__.py [hunk 8]
```diff
@@ -1,6 +1,9 @@
+from .bpe import BPEConfig, BPETokenizer
+from .sentencepiece_bpe import SentencePieceBPEConfig, SentencePieceBPETokenizer
+from .sentencepiece_unigram import (
+    SentencePieceUnigramConfig,
+    SentencePieceUnigramTokenizer,
+)
 from .tokenizer import Tokenizer, TokenizerConfig
```
Comment (arxyzan) [inline hezar/preprocessors/tokenizers/__init__.py:7]: @HamedBabaei Tokenizer is the base class here and must come first.

File: hezar/preprocessors/preprocessor.py (hunk 5 above)
Comment (HamedBabaei) [reply hezar/preprocessors/preprocessor.py]: After turning `isort` backend to `black`, the current method keeps such cases untouched. I will try to find them and roll them back as much as I could do!

File: hezar/preprocessors/tokenizers/__init__.py (hunk 8 above)
Comment (HamedBabaei) [reply hezar/preprocessors/tokenizers/__init__.py:7]: Done!

**Comment by arxyzan** [comment]
@HamedBabaei jan, can you please pull the recent changes to `hezar/data/datasets/dataset.py` to resolve the recent conflict?

**Comment by arxyzan** [comment]
@HamedBabaei By the way, did you verify that all black, ruff, isort, etc rules do not conflict with each other? It's best to configure all of them to follow a common practice so that devs do not get warnings all the time. 
Since the beginning I used ruff and I'm not much familiar with these tools after all. Just make sure we're consistent in all places and for the future.

**Comment by HamedBabaei** [comment]
> @HamedBabaei By the way, did you verify that all black, ruff, isort, etc rules do not conflict with each other? It's best to configure all of them to follow a common practice so that devs do not get warnings all the time. Since the beginning I used ruff and I'm not much familiar with these tools after all. Just make sure we're consistent in all places and for the future.

Yes. To the best of my knowledge, the black and isort had some conflicts so I just turned `isort` backend to `black` and ignored `black` in pre-commits. Sometimes it is hard to work with both of them due to some conflicts that they may cause in some points of development. So personally, I consider one of them with flake configurations to control my changes. I have been using both of them but came to the conclusion that for the start it would be better to consider one of them and move forward, and by the time of requiring more stable signatures in coding adding more configurations would hurt the development! 

I didn't also use ruff, I just used `pre-commit` itself because of some `hooks` that were needed in projects. And it is a matter of personal preference to work which of them.

**Comment by arxyzan** [comment]
> > @HamedBabaei By the way, did you verify that all black, ruff, isort, etc rules do not conflict with each other? It's best to configure all of them to follow a common practice so that devs do not get warnings all the time. Since the beginning I used ruff and I'm not much familiar with these tools after all. Just make sure we're consistent in all places and for the future.
> 
> Yes. To the best of my knowledge, the black and isort had some conflicts so I just turned `isort` backend to `black` and ignored `black` in pre-commits. Sometimes it is hard to work with both of them due to some conflicts that they may cause in some points of development. So personally, I consider one of them with flake configurations to control my changes. I have been using both of them but came to the conclusion that for the start it would be better to consider one of them and move forward, and by the time of requiring more stable signatures in coding adding more configurations would hurt the development!
> 
> I didn't also use ruff, I just used `pre-commit` itself because of some `hooks` that were needed in projects. And it is a matter of personal preference to work which of them.

I previously researched black and ruff and came to the conclusion that ruff is overall both faster (written with Rust) and has better default rules so we stuck with that and honestly I still prefer it as it's been pretty fine for me.

**Comment by arxyzan** [comment]
@HamedBabaei Plus, some styling rules in Black are considered obsolete by the community for a while.

**Comment by HamedBabaei** [comment]
> @HamedBabaei Plus, some styling rules in Black are considered obsolete by the community for a while.

Yes, I agree, and because of this I just ignored it, but in some other projects that I have I am using both of them https://github.com/TIBHannover/orkg-nlp-pypi

**Comment by arxyzan** [comment]
@HamedBabaei Most modern Python tools nowadays use ruff like Hugging Face, LangChain, LlamaIndex, FastAPI, PyTorch, ZenML, etc. That's why we choosed ruff over other tools.
Let me know what you think.

**Comment by HamedBabaei** [comment]
> @HamedBabaei Most modern Python tools nowadays use ruff like Hugging Face, LangChain, LlamaIndex, FastAPI, PyTorch, ZenML, etc. That's why we choosed ruff over other tools. Let me know what you think.

Yes, but it depends on how they use it!

For LangChain they are using it with the following remarks from ruff :
```
[tool.ruff]
select = [
  "E",  # pycodestyle
  "F",  # pyflakes
  "I",  # isort
]
```

or in transformers:

```
[tool.ruff]
ignore = ["C901", "E501", "E741"]
select = ["C", "E", "F", "I", "W"]
line-length = 119
```
I applied the same remarks that you have in https://github.com/hezarai/hezar/blob/main/pyproject.toml to the `flake8` : https://github.com/HamedBabaei/hezar/blob/dev/.flake8

but considering transformers and langchain, I didn't see whatever they applied `isort` or `black`, only it was just `ruff` in `.tmol` file, but not in the specific pre-commit-config.yml file that allows users to fix their issue while committing! 

overall the `ruff-pre-commit` base is the `pre-commit` but if you want to move `pre-commit` to `ruff-pre-commit`  it shouldn't be a problem! And we might ignore the followings from  https://github.com/hezarai/hezar/pull/104/files#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9 (if they are not available in ruff)

```
    -   id: check-merge-conflict
    -   id: end-of-file-fixer
    -   id: trailing-whitespace
    -   id: detect-private-key
    -   id: check-added-large-files
    -   id: check-byte-order-marker
    -   id: mixed-line-ending
 ```

I look forward to continuing the discussion! 

Thanks for your time!

**Comment by arxyzan** [comment]
@HamedBabaei Based on the current changes to the files, it seems that there's not much different rules between your configuration and ours. So let's just finalize the current changes and merge.

File: hezar/metrics/__init__.py [hunk 9]
```diff
@@ -1,7 +1,7 @@
-from .f1 import F1, F1Config
 from .accuracy import Accuracy, AccuracyConfig
 from .bleu import BLEU, BLEUConfig
 from .cer import CER, CERConfig
+from .f1 import F1, F1Config
 from .metric import Metric
```
Comment (arxyzan) [inline hezar/metrics/__init__.py:5]: @HamedBabaei Forgot to mention this one. Metric must be on top.

**Comment by HamedBabaei** [comment]
> @HamedBabaei Based on the current changes to the files, it seems that there's not much different rules between your configuration and ours. So let's just finalize the current changes and merge.

Great! thanks for the feedback

File: hezar/metrics/__init__.py (hunk 9 above)
Comment (HamedBabaei) [reply hezar/metrics/__init__.py]: done!

**Comment by arxyzan** [comment]
@HamedBabaei Sorry for the inconvenience Hamed jan, but I noticed that the `lines-after-imports=2` in `.flake8` does not work and all codes have 1 line after import. What am I missing here?

**Comment by HamedBabaei** [comment]
> @HamedBabaei Sorry for the inconvenience Hamed jan, but I noticed that the `lines-after-imports=2` in `.flake8` does not work and all codes have 1 line after import. What am I missing here?

@arxyzan jan, Awesome catch, thanks for doing that! I noticed that flake after 2021 doesn't support this functionality, so I just added it to the `isort` args. Now it works fine!
